### PR TITLE
Enable multiple outputs in BIP70 payment protocol requests

### PR DESCRIPTION
**Pull request by vinarmani**
Many tests in this project have been broken for a long time. Fixed payment protocol tests.

### CODE REVIEW

**Comment by vinarmani** [comment]
BIP70 payment protocol supports multiple outputs. This module had been simply grabbing the output at index 0 and only sending that forward. This is non-standard behavior. I have left the previous behavior for backward compatibility, but also added an "outs" property to the result object that is an array of outputs (address and amount).

**Comment by vinarmani** [comment]
If the below pull request:

https://github.com/Bitcoin-com/bitcore-payment-protocol/pull/1

Id done in that order then the dependency can remain pointing at the modules in the Bitcoin.com github repo for bitcore-payment-protocol.

File: lib/paypro.js [hunk 1]
```diff
@@ -124,25 +124,29 @@ PayPro.get = function(opts, cb) {
...
+    
+    var network = pd.get('network') == 'test' ? 'testnet' : 'livenet';
 
     var outputs = pd.get('outputs');
-    if (outputs.length > 1)
-      return cb(new Error('Payment Protocol Error: Requests with more that one output are not supported'))
+    var outs = outputs.map(function(output){
 
-    var output = outputs[0];
+      let amount = output.get('amount').toNumber();
```
Comment (brendoncoin) [inline lib/paypro.js]: Replace let with var, as early devices we run on do not support let.

File: lib/paypro.js [hunk 2]
```diff
@@ -166,8 +170,9 @@ PayPro.get = function(opts, cb) {
       memo: pd.get('memo'),
       time: pd.get('time'),
       merchant_data: md,
-      toAddress: addr.toString(),
-      amount: amount,
+      toAddress: outs[0].addr,
```
Comment (brendoncoin) [inline lib/paypro.js:174]: This assumes that the first address will always be the "primary" one? Perhaps a check should be made to ensure this is selecting the address that is receiving the most? Does this `addr` need a `toString()`?

File: package.json [hunk 3]
```diff
@@ -29,7 +29,7 @@
     "bitcore-lib": "0.14.0",
     "bitcore-lib-cash": "https://github.com/Bitcoin-com/bitcore-lib-cash.git",
     "bitcore-mnemonic": "1.2.2",
-    "bitcore-payment-protocol": "https://github.com/Bitcoin-com/bitcore-payment-protocol.git",
```
Comment (brendoncoin) [inline package.json]: Needs to link to the updated bitcoin.com repo once the changes have been made to it.

File: test/paypro.js [hunk 4]
```diff
@@ -56,6 +56,7 @@ describe('paypro', function() {
   });
 
   it('Make a PP request with browser', function(done) {
+    xhr.status=200;
```
Comment (brendoncoin) [inline test/paypro.js]: Needs a space either side of the = operator.

File: lib/paypro.js [hunk 5]
```diff
@@ -124,25 +124,29 @@ PayPro.get = function(opts, cb) {
...
+      let addr = new bitcore.Address.fromScript(new bitcore.Script(scriptBuf), network);
+      return {addr: addr.toString(), amount: amount}
+    });
 
-    // NOTE: For some reason output.script.buffer
-    // is only an ArrayBuffer
-    var buffer = new Buffer(new Uint8Array(output.get('script').buffer));
-    var scriptBuf = buffer.slice(offset, limit);
-    var addr = new bitcore.Address.fromScript(new bitcore.Script(scriptBuf), network);
+    var totalAmount = outs.reduce(function (a, b) {
```
Comment (brendoncoin) [inline lib/paypro.js]: Name the anonymous function.

Comment (vinarmani) [reply lib/paypro.js]: are you referring to the callback function being passed as an argument to reduce?

File: lib/paypro.js (hunk 2 above)
Comment (vinarmani) [reply lib/paypro.js:174]: This pull request enables multiple outputs. Previously, the first index was the only one selected. Explicitly selecting that first index for toAddress maintains backwards compatibility. A new property ("outs") has been added to this object that is an array of outputs. That is what is used in my upgrades to the wallet, but maintaining the toAddress property means that this new library is still backwards compatible with all older wallet versions.

Comment (vinarmani) [reply lib/paypro.js:174]: The toString() is now done in line 144

Comment (brendoncoin) [reply lib/paypro.js:174]: The backwards compatibility angle makes sense, thanks for the explanation.

File: lib/paypro.js (hunk 5 above)
Comment (brendoncoin) [reply lib/paypro.js]: That's correct. The existing code doesn't always have that, but that's what we are doing going forward. It helps by expressing your intention and also provides better information in the call stack when debugging.

Comment (vinarmani) [reply lib/paypro.js]: Changes made

File: package.json (hunk 3 above)
Comment (vinarmani) [reply package.json]: Ready to do that whenever the PR is done on bitcore-payment-protocol

File: test/paypro.js (hunk 4 above)
Comment (vinarmani) [reply test/paypro.js]: Done

**Comment by vinarmani** [comment]
Requested changes made. Passes unit test

File: lib/paypro.js [hunk 6]
```diff
@@ -124,25 +124,34 @@ PayPro.get = function(opts, cb) {
...
+    };
+
+    var amountReduce = function (a, b) {
+      return b['amount'] == null ? a : a + b['amount'];
+    };
 
-    // We love payment protocol
-    var offset = output.get('script').offset;
-    var limit = output.get('script').limit;
+    var outs = outputs.map(mapOutputs);
```
Comment (jbdtky) [inline lib/paypro.js]: `mapOutputs` should be replaced by the function, used only once. It removed readability.

File: lib/paypro.js [hunk 7]
```diff
@@ -124,25 +124,34 @@ PayPro.get = function(opts, cb) {
...
 
-    // NOTE: For some reason output.script.buffer
-    // is only an ArrayBuffer
-    var buffer = new Buffer(new Uint8Array(output.get('script').buffer));
-    var scriptBuf = buffer.slice(offset, limit);
-    var addr = new bitcore.Address.fromScript(new bitcore.Script(scriptBuf), network);
+    var totalAmount = outs.reduce(amountReduce, 0);
```
Comment (jbdtky) [inline lib/paypro.js]: `amountReduce` should be replaced by the function, used only once and only 1 line. It removed readability.

File: lib/paypro.js (hunk 6 above)
Comment (jbdtky) [reply lib/paypro.js]: To follow our guidelines and avoid any crash with old javascript engine.
```javascript
var outs = outputs.map(function onMap(output) {
 // your code...
});
```

File: lib/paypro.js (hunk 7 above)
Comment (jbdtky) [reply lib/paypro.js]: To follow our guidelines and avoid any crash with old javascript engine.
```javascript
var totalAmount = outs.reduce(function onReduce(a, b) {
 // your code...
});
```

File: test/testdata.js [hunk 8]
```diff
@@ -86,6 +86,8 @@ var payProData = {
   toAddress: 'mjfjcbuYwBUdEyq2m7AezjCAR4etUBqyiE',
   amount: 404500,
   network: 'testnet',
+  outs:
+   [ { addr: 'mjfjcbuYwBUdEyq2m7AezjCAR4etUBqyiE', amount: 404500 } ],
```
Comment (jbdtky) [inline test/testdata.js:90]: Please add 1 test with more than 1 output.

**Review by jbdtky — CHANGES_REQUESTED** [review]

File: lib/paypro.js (hunk 6 above)
Comment (vinarmani) [reply lib/paypro.js]: Done

File: lib/paypro.js (hunk 7 above)
Comment (vinarmani) [reply lib/paypro.js]: Done

File: test/testdata.js (hunk 8 above)
Comment (vinarmani) [reply test/testdata.js:90]: These tests are old and defunct. They were put in place by Bitpay, who had uploaded binary files matching the data in testdata. There are no binary files with multiple outputs to match against. I would need to redo basically all of the tests to do this.

**Pull request merged**